sources/web/datalab/static/idle-timeout.js

```diff
@@ -0,0 +1,264 @@
+  const updateTimeoutInfoInterval = 1000; // Update display every second while the user is watching.
+  const updateTimeoutInfoNoDisplayInterval = 5 * 1000; // Keep tabs on timeout info even when not being displayed.
+  const queryTimeoutInfoInterval = 10 * 1000; // Only send a query every ten seconds.
+  let timeoutInfoTimeout;     // Result of setTimeout for calling updateTimeoutInfo
+  let timeoutInfo = {};
+  let lastUpdateTimeoutTime = 0;
+
+  let showDebugLog = false;   // Set to true in the dev console to get log messages during development.
+  function _debugLog(msg) {
+    if (showDebugLog) {
+      console.log(msg);
+    }
+  }
+
+  // Sets event handlers to deal with kernel-busy heartbeats.
+  function setupKernelBusyHeartbeat() {
+    events.on('kernel_busy.Kernel', function() {
+      _debugLog('== kernel_busy event received');
+      _runBusyTimer();
+    });
+    events.on('kernel_idle.Kernel', function() {
+      _debugLog('== kernel_idle event received');
+      _clearBusyTimer();
+      _maybeResetIdleTimeout();
+    });
+  }
+
+  // Sets an event handler to deal with kernel-disconnected events.
+  function setupDisconnectHandler() {
+    events.on('kernel_connection_failed.Kernel', function() {
+      _debugLog('== Disconnected; timeoutInfo:');
+      _debugLog(timeoutInfo);
+      _alertIfTimedOut();
+    });
+  }
+
+  function _alertIfTimedOut() {
+    if (timeoutInfo.enabled && timeoutInfo.idleTimeoutSeconds > 0) {
+      const secondsRemaining = Math.floor((timeoutInfo.expirationTime - Date.now()) / 1000);
+      if (secondsRemaining <=1 ) {
+        // If there are other handlers, let them run first.
+        const delayToLetOtherHandlersRun = 2 * 1000;
+        window.setTimeout(_alertIdleShutdown, delayToLetOtherHandlersRun);
+      }
+    }
+    timeoutInfo.enabled = false;  // Make sure we don't get a false indication later.
+  }
+
+  // Alerts the user that there was a shutdown due to idle time exceeded.
+  function _alertIdleShutdown() {
+    _debugLog('== Idle timeout exceeded');
+    const shutdownMsg = ('The datalab server shut down after exceeding the idle timeout of ' +
+        _secondsToString(timeoutInfo.idleTimeoutSeconds));
+    dialog.modal({
+      title: 'Idle timeout exceeded',
+      body: shutdownMsg,
+      buttons: { 'Close': {} },
+    });
+    timeoutInfo.enabled = false;  // Don't show this message more than once.
+  }
+
+  // Initializes the idle-timeout mechanism.
+  function setupTimeoutTimer() {
+    updateTimeoutInfo(undefined);
+  }
+
+  // Updates the display of timeout info in the dropdown menu.
+  // Assumes global timeoutInfo has been set.
+  function _updateTimeoutDisplay(dropdown) {
+    if (!dropdown) {
+      _debugLog('== updateTimeoutDisplay no dropdown');
+      return;
+    }
+    _debugLog('== updating timeout display');
+    _debugLog(dropdown);
+    if (timeoutInfo.enabled) {
+      let secondsRemaining = Math.floor((timeoutInfo.expirationTime - Date.now()) / 1000);
```

Comment:
I really don't think this should be in seconds, because of the fact that the client code isn't always in sync with the server state. the user will likely see the timer go down second by second, then jump up when the server resets and the client polls the new state.
minutes makes a lot more sense to me.